src/mount/writedata.cc

```diff
@@ -94,12 +92,63 @@ struct inodedata {
 			  flushwaiting(0),
 			  writewaiting(0),
 			  lcnt(0),
-			  trycnt(0),
-			  inqueue(false),
-			  minimumBlocksToWrite(1),
-			  alterations_in_chain(),
 			  next(nullptr),
-			  workerWaitingForData(false) {
+			  chunksDataFirst(nullptr),
+			  totalCachedBlocks(0) {
+	}
+
+	/*! Check if inode requires flushing all its data chain to chunkservers.
+	 *
+	 * Returns true if anyone requested to flush the data by calling write_data_flush
+	 * or write_data_flush_inode or the data in data chain is too old to keep it longer in
+	 * our buffers. If this function returns false, we write only full stripes from data
+	 * chain to chunkservers.
+	 * inodeLock: LOCKED
+	 */
+	bool requiresFlushing() {
+		return (flushwaiting > 0 ||
+		        lastWriteToDataChain.elapsed_ms() >=
+		            kMaximumTimeInDataChainSinceLastWrite_ms ||
+		        lastWriteToChunkservers.elapsed_ms() >=
+		            kMaximumTimeInDataChainSinceLastFlush_ms);
+	}
+
+private:
+	/*! Limit for \p lastWriteToChunkservers after which we force a flush.
+	 *
+	 * Maximum time for data to be kept in data chain waiting for collecting a full stripe.
+	 */
+	static const uint32_t kMaximumTimeInDataChainSinceLastFlush_ms = 15000;
+
+	/*! Limit for \p lastWriteToDataChain after which we force a flush.
+	 *
+	 * Maximum time for data to be kept in data chain waiting for collecting a full stripe
+	 * if no new data is written into the data chain
+	 */
+	static const uint32_t kMaximumTimeInDataChainSinceLastWrite_ms = 5000;
+};
+
+struct ChunkData {
+	uint32_t chunkIndex;
+	uint16_t tryCounter;
+	int newDataInChainPipe[2];
+	uint32_t minimumBlocksToWrite;
+	std::list<WriteCacheBlock> dataChain;
+	bool inQueue; // true it this chunk is waiting in one of the queues or is being processed
+	bool workerWaitingForData;
+	ChunkData *next; // for iterating through the chunkdata of an inode
+	std::unique_ptr<WriteChunkLocator> locator;
+	std::atomic<int> beingWritten;
```

Comment:
Well, it could be changed as you propose because we are not allowing more than one write on an inode at the same time (thanks to the fileinfo's lock), but what if we change it and allow multiple writes somehow later. My idea about beingWritten is like "how many times are you concurrently being written?".

I did not apply it in this first set of review applies.

Reply:
So the name could be changed to writesCount or similar?

Reply:
Yes, and it would look better. Applied.